### PR TITLE
notify dashboard migration fail

### PR DESCRIPTION
**Pull request by mtanda**
When updating Grafana 2.x to 3.x, sometimes dashboard migration failed, and empty dashboard shown.
I think the error should notify to user, but current, there is no error dialog.

@torkelo Would you check this and include this to next release?

### CODE REVIEW

**Comment by torkelo** [comment]
Removed migration from 1.x so it's a bit late for this fix

**Comment by torkelo** [comment]
Ops read the issue text wrong :)

**Comment by torkelo** [comment]
So maybe this exception handling should be in dashboard controller? What is the exception?

**Comment by mtanda** [comment]
@torkelo The exception is due to mismatched json format of dashboard.

For example, I set dashboard json like following by scripted dashboard in 2.x.
```
"grid": {
  "leftMin": 0,
  "leftMax": 100
},
"y_formats": [
  "percent"
},
```

I change this json for 3.x, but it failed to migration.
```
"yaxes": [
  {
    "show": true,
    "min": 0,
    "max": 100,
    "format": "percent"
  }
]
```

I need to set like following.
```
"yaxes": [
  {
    "show": true,
    "min": 0,
    "max": 100,
    "format": "percent"
  },
  {
    "show": false,
    "min": null,
    "max": null,
    "format": "short"
  }
],
```

**Comment by mtanda** [comment]
When I see empty dashboard, I can't imagine why it doesn't work.
I'm not sure this is rare case or not, but I think preparing error notification is very important, because the migration process could receive wrong format dashboard json.

**Comment by torkelo** [comment]
of course, was just wondering what the cause for the exception was and if it should be fixed. 

I think instead of using alertSrv throw an error that can be handled more centrally (in setupDashboard)

**Comment by mtanda** [comment]
Sorry, I can't replicate the issue now.
But, I remember error is caused when accessing array, but array is undefined or null.
https://github.com/grafana/grafana/blob/0d4c76a0291032a8a9b3ad96f353f150647ceafc/public/app/features/dashboard/dashboardSrv.js#L442

**Comment by mtanda** [comment]
move to exception handling code to dashboard controller.

File: public/app/features/dashboard/dashboard_ctrl.ts [hunk 1]
```diff
@@ -29,11 +30,15 @@ export class DashboardCtrl {
...
-        // init services
-        timeSrv.init(dashboard);
+        try {
+          var dashboard = dashboardSrv.create(data.dashboard, data.meta);
+          dashboardSrv.setCurrent(dashboard);
+
+          // init services
+          timeSrv.init(dashboard);
+        } catch (err) {
+          $scope.appEvent("alert-error", ['Dashboard init failed', err.message]);
```
Comment (torkelo) [inline public/app/features/dashboard/dashboard_ctrl.ts]: this does not look right will just continue down the function and cause other errors? Maybe combine this exception handler with the one below ?

Comment (mtanda) [inline public/app/features/dashboard/dashboard_ctrl.ts]: Ah, yes. should abort when failed.

Comment (mtanda) [inline public/app/features/dashboard/dashboard_ctrl.ts]: But, the exception handler you mentioned is Promise's one.
I don't know well about it, is it possible to combine these exception handler?

https://github.com/mtanda/grafana/blob/44b2a292eb1e467b07b2941246ab8e0e153dffd3/public/app/features/dashboard/dashboard_ctrl.ts#L45

**Comment by torkelo** [comment]
refactored this a bit, separate fatal and non fatal dashboard init errors, on fatal errors it will try to load fallback dashboard (so you are not staring at black screen), non fatal errors like template variable loading errors should not stop dashboard from loading so they are considered non fatal.

https://github.com/grafana/grafana/commit/b4493155fc6aaf1273d3699312a153dada01ec07

**Comment by torkelo** [comment]
linked to wrong commit https://github.com/grafana/grafana/commit/90880231169b9890b28cdcd7b5098f6744f367d0

**Pull request merged**